add distinctions re: local- vs remote-eval (#276)
open-feature/docs.openfeature.dev: blog/06-02-2023-catering-to-the-client-side.md

In contrast, with a client-side app all feature flagging decisions are made in the context of the same user - the user interacting with the client-side app - and so the evaluation context is relatively static. There are cases where evaluation context will change within a client-side app - when a user logs in, for example - but by and large with client-side code we can treat feature flag evaluation context as something that is fixed (while still providing mechanisms to update it).

### The network is slow

With server-side flags, we can assume that evaluating a feature flag is a _relatively_ fast operation. With some systems the flagging rulesets[^1] live right next to where a flagging decision is needed, with flag evaluation happening either within the same process or in some sort of sidecar process. In this *local evaluation* model every flagging decision is a very fast operation. For frameworks that use a *remote evaluation* model, a flagging decision is still just a quick service call - akin to making a DB query or calling a remote cache.

> [^1]: by "rulesets" I mean the set of feature flagging rule configurations that define how flagging decisions should be made for each feature flag: "enable red_checkout_button flag for 50% of users", "only enable new_reco_algorithm for users in the 'internal_testers' group", etc. Rulesets plus evaluation context are the two inputs that fully define the output for any flagging decision.

Reviewer: You may want to consider using an [Admonition](https://docusaurus.io/docs/markdown-features/admonitions) or [quote](https://docusaurus.io/docs/markdown-features#quotes).

Reviewer: Ignore this... your way is better 😄

> This situation is quite different with client-side flags. Remote flag evaluation now requires a trip across the internet, and we have to anticipate such a service call to be slow, particularly if our users are behind a spotty internet connection. In fact, with a native mobile app we have to handle a fully disconnected client. Even with the local evalution model, we still have to deal with our fact that the source of truth for our flagging ruleset is on the other side of a potentially high-latency network.

Reviewer: ```suggestion
This situation is quite different with client-side flags. Remote flag evaluation now requires a trip across the internet, and we have to anticipate such a service call to be slow, particularly if our users are behind a spotty internet connection. In fact, with a native mobile app we have to handle a fully disconnected client. Even with the local evaluation model, we still have to deal with the fact that the source of truth for our flagging ruleset is on the other side of a potentially high-latency network.
```

### Eager evaluation for remote-evaluated systems

We can see that the network presents challenges for client-side apps using a remote evaluation model for flagging decisions. But we've also seen that the inputs into that flag evaluation - the evaluation context - are fairly static for client-side apps, and that means the _results_ of flag evaluation are fairly static too.

> How do we handle an expensive operation with fairly static results? We add caching! And that's what most client-side feature flagging frameworks do.

Reviewer: ```suggestion
How do we handle an expensive operation with fairly static results? We add caching! And that's what many client-side feature flagging frameworks do.
```

Put another way, with client-side feature flagging we can separate flag [**evaluation**](/docs/specification/glossary#evaluating-flag-values) - passing an evaluation context through a set of rules in order to determine a flagging decision - from flag [**resolution**](/docs/specification/glossary#resolving-flag-values) - getting the flagging decision for a specific feature flag.

## Keeping rulesets in sync for local-evaluated systems 

> Flagging frameworks that use a local evaluation model don't have to contend with network calls for every evaluation, but they still need to keep their local rulesets up to date and allow the client-side app to respond to changes in those ruleset. Again, this means using caches to keep the most recent ruleset available. It also means that the flagging framework needs some eventing or callback mechanism to inform application code that the ruleset has changed and flagging decision needs to be re-evaluated. 

Reviewer: ```suggestion
Flagging frameworks that use a local evaluation model don't have to contend with network calls for every evaluation, but they still need to keep their local rulesets up to date and allow the client-side app to respond to changes in those ruleset. Again, this means using caches to keep the most recent ruleset available. It also means that the flagging framework may have an eventing or callback mechanism to inform application code that the ruleset has changed and that current flag values may be stale.
```
I'm suggesting a more passive tone here since eventing is a relatively advanced feature that isn't available in every tool.

### Keeping rulesets in sync for local-evaluated systems

Flagging frameworks that use a local evaluation model don't have to contend with network calls for every evaluation, but they still need to keep their local rulesets up to date and allow the client-side app to respond to changes in those ruleset. Again, this means using caches to keep the most recent ruleset available. It also means that the flagging framework may have an eventing or callback mechanism to inform application code that the ruleset has changed and that current flag values may be stale.

```mermaid
sequenceDiagram
  actor app
  participant sdk as feature flag SDK
  participant cache as local ruleset
  participant provider as flag service
  app->>+sdk: boot
  sdk->>+provider: get ruleset from last known state
  provider-->>-sdk: ruleset
  sdk->>+cache: Store ruleset
  cache->>-sdk: .
  sdk->>-app: .
  app->>+sdk: operation that needs a flagging decision
  sdk->>+cache: flag evaluation
  cache->>-sdk: locally evaluated flag value
  sdk->>-app: flag value 

>   note left of provider: some time later, flagging rules have been updated

Reviewer: You should consider including this in the eager evaluation diagram as well. Both techniques can support cache busting.
